src/main/java/com/google/sps/servlets/BlobServlet.java

```diff
@@ -33,59 +41,203 @@
   public void doPost(HttpServletRequest request, HttpServletResponse response)
       throws IOException {
     UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
+    // Must be logged in
     if (!userService.isUserLoggedIn()) {
-      response.sendRedirect("/imgupload.html");
+      throw new IOException("User must be logged in.");
+    }
+
+    // Mode is a required parameter
+    String mode = request.getParameter("mode");
+    if (mode == null || (mode != "create" && mode != "update")) {
+      throw new IOException("Invalid mode.");
+    }
+
+    // Check if working with image or mask
+    String parentImg = request.getParameter("parent-img");
+    Boolean isMask = false;
+    if (parentImg != null) {
+      isMask = true;
+    }
+
+    // Get the image name entered by the user
+    String imgName = request.getParameter("img-name");
+    if (imgName == null) {
+      throw new IOException("Image name must be provided.");
+    }
+
+    String projId = request.getParameter("proj-id");
+    Key projKey = KeyFactory.stringToKey(projId);
+    Entity projEntity = new Entity("Project");
+    try {
+      projEntity = datastore.get(projKey);
+    } catch (Exception e) {
+      response.sendRedirect("/");
       return;
     }
+    String uid = userService.getCurrentUser().getUserId();
+    Key assetParentKey = projKey;
+    Entity imgEntity = new Entity("Image", projKey);
 
-    // Get the name entered by the user
-    String name = request.getParameter("name");
+    // Must be an owner or editor
+    Query projQuery = new Query("Project");
+
+    Filter ownEditFilter = new CompositeFilter(
```

Comment:
No change request here, but can you explain to me what this line does?

Reply:
It's creating a new filter for the Datastore query. It's searching for projects with the given ID and with the User as either owner or editor.